### PR TITLE
[deckhouse-controller] add a convenient way of resolving which deckhouse pod is current leader

### PR DESCRIPTION
**Pull request by juev**
## Description

We need to add a new label to the leader

## Why do we need it, and what problem does it solve?

Previously, only one pod deckhouse was launched during deployment, and when the remote shell was launched, all commands were correctly processed.

Now it is possible to launch several pods for the deckhouse application, one is called the leader, the rest are in standby mode. As a result, when launching a remote shell, we find ourselves in an arbitrary pod, and the commands do not provide the necessary information about the system.

## What is the expected result?

When using a new label, you can address the launch of a remote shell in the pod leader.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: feature
summary: add a convenient way of resolving which deckhouse pod is current leader
impact_level: default
```

Closes: #8403

### CODE REVIEW

File: modules/002-deckhouse/templates/deckhouse-leader-service.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: deckhouse-leader
+  namespace: d8-system
+  {{- include "helm_lib_module_labels" (list . (dict "app" "deckhouse")) | nindent 2 }}
+spec:
```
Comment (miklezzzz) [inline modules/002-deckhouse/templates/deckhouse-leader-service.yaml:8]:
```suggestion
spec:
  publishNotReadyAddresses: true
```
sorry, my bad, I forgot to mention this parameter at the beginning

File: modules/002-deckhouse/hooks/set_deckhouse_leader_pod.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	return nil
+}
+
+func emptyOrNil(b bool) any {
+	if b {
+		return ""
+	}
+
+	return nil
+}
```
Comment (miklezzzz) [inline modules/002-deckhouse/hooks/set_deckhouse_leader_pod.go:75]: I tested in a cluster and it takes Deckhouse substantial amount of time to reach the point when the hook is evoked, could you please move the hook as it is to ./global-hooks/ ?

File: modules/002-deckhouse/templates/deckhouse-leader-service.yaml (hunk 1 above)
Comment (juev) [reply modules/002-deckhouse/templates/deckhouse-leader-service.yaml:8]: I added the parameter

File: modules/002-deckhouse/hooks/set_deckhouse_leader_pod.go (hunk 2 above)
Comment (juev) [reply modules/002-deckhouse/hooks/set_deckhouse_leader_pod.go]: I moved the files to global-hooks

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Review by miklezzzz — APPROVED** [review]
works as a charm

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: global-hooks/set_deckhouse_leader_pod.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+		input.PatchCollector.MergePatch(patch, "v1", "Pod", d8Namespace, pod.Name)
+	}
+
+	return nil
+}
+
+func trueOrNil(b bool) any {
+	if b {
+		return true
```
Comment (yalosev) [inline global-hooks/set_deckhouse_leader_pod.go:67]:
```suggestion
		return "true"
```

Label could be only `string` or nil. So boolean will fail here

File: modules/002-deckhouse/templates/deckhouse-leader-service.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  publishNotReadyAddresses: true
+  selector:
+    app: deckhouse
+    leader: true
```
Comment (yalosev) [inline modules/002-deckhouse/templates/deckhouse-leader-service.yaml:12]:
```suggestion
    leader: "true"
```

Comment (yalosev) [reply modules/002-deckhouse/templates/deckhouse-leader-service.yaml:12]: Just to add some safety. `deckhouse` label could be interpreted only as string, while the `leader` label could be interpreted as boolean. You will get an error:
```
 Service in version "v1" cannot be handled as a Service: json: cannot unmarshal bool into Go struct field ServiceSpec.spec.selector of type string
 ```

Comment (yalosev) [reply modules/002-deckhouse/templates/deckhouse-leader-service.yaml:12]: For example, you have to to like this with some values, like `controllerVersion`, which is string, but if you set `1.10` there, kubernetes will interpret it like a float

Comment (yalosev) [reply modules/002-deckhouse/templates/deckhouse-leader-service.yaml]: At the same time, if we do `kubectl get pods -l leader=true`, kubectl will case this values to the string

File: global-hooks/set_deckhouse_leader_pod.go (hunk 3 above)
Comment (juev) [reply global-hooks/set_deckhouse_leader_pod.go]: Fixed

File: modules/002-deckhouse/templates/deckhouse-leader-service.yaml (hunk 4 above)
Comment (juev) [reply modules/002-deckhouse/templates/deckhouse-leader-service.yaml]: Thank you! 
Now it has become clearer.

**Comment by yalosev** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.28 9710263297 9710263297-1-con-1-28 /sys/deckhouse-oss/install:pr8720 redos@158.160.114.70

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Pull request merged**